README.md
# Sudiyi Open Platform SDK for PHP

[![Latest Stable Version](https://poser.pugx.org/sudiyi/sudiyi-open-php-sdk/version)](https://packagist.org/packages/sudiyi/sudiyi-open-php-sdk)
[![Build Status](https://travis-ci.org/sudiyi/sudiyi-open-php-sdk.svg?branch=master)](https://travis-ci.org/sudiyi/sudiyi-open-php-sdk)
[![Coverage Status](https://coveralls.io/repos/github/sudiyi/sudiyi-open-php-sdk/badge.svg?branch=master)](https://coveralls.io/github/sudiyi/sudiyi-open-php-sdk?branch=master)

## 概述

速递易开放平台是速递易为企业或商户开通的速递易箱格接口，无缝对接PC网站、微信公众号、APP，帮助商家提升配送效率，降低配送成本，解决快件最后100米难题。

## 运行环境
- PHP 5.3+
- cURL extension

## 安装方法

1. 如果您通过composer管理您的项目依赖，可以在你的项目根目录运行：

        $ composer require sudiyi/sudiyi-open-php-sdk

   或者在你的`composer.json`中声明对速递易开放平台 SDK 的依赖：

        "require": {
            "sudiyi/sudiyi-open-php-sdk": "~1.0"
        }

   然后通过`composer install`安装依赖。

2. 下载SDK源码，在您的代码中引入 SDK 目录下的`autoload.php`文件：

        require_once '/path/to/sudiyi-open-php-sdk/autoload.php';

## SDK 主要目录结构

```
|-- example
|   |-- api_use_demo.php        案例程序代码
|   `-- notify_demo.php         商户回调处理
|-- src
|   |-- Core
|   |    |-- Base.php           基础类
|   |    |-- Config.php         常量配置
|   |    |-- Http.php           Http请求处理类
|   |    `-- SdyException.php   异常类
|   `-- Api.php                 接口封装类
|-- tests/                      测试文件
|-- autoload.php                PSR-4 自动加载                     
`-- composer.json
```

## 快速使用

`Sudiyi\Open\Api` 是对速递易开方平台的接口封装,用户通过 Api 的实例来执行各种操作

### 初始化

```php
require_once'/path/to/sudiyi-open-php-sdk/autoload.php';

use Sudiyi\Open\Api;

$partner_id = "<您从速递易开放平台获得的 PartnerId>";
$partner_key = "<您从速递易开放平台获得的 PartnerKey>";

$sdyClient = new Api($partner_id, $partner_key);
```

### 调用接口

```php
$data = array(
    'device_id'         => 1000149,                         // 速递易设备ID
    'box_type'          => 2,                               // 箱格类型 0:大箱 1:中箱 2:小箱 3:冰箱
    'notify_url'        => 'http://your.url/to/notify.php', // 回调第三方系统的地址
    'auto_upgd'         => true,                            // 选填，是否可以自动升箱
    'sender_name'       => 'Stefans',                       // 选填，投递人姓名
    'sender_mobile'     => '18612345678',                   // 选填，投递人手机
    'order_no'          => '2015000001',                    // 第三方订单号,即自己平台的订单号
    'consignee_name'    => 'demon',                         // 收货人姓名
    'consignee_mobile'  => '15887654321',                   // 收货人手机号
    'payment'           => 0,                               // 选填，到付金额，如果不是到付件则为0
    'pay_type'          => 0,                               // 选填，超期付费方式 0:快递柜付费 1:从商户账户扣除
    'duration'          => 60                               // 选填，预约时长，单位分钟，默认240分钟
);

// 预约速递易箱格
$result = $sdyClient->resv($data);
echo "预约成功: \n";
var_dump($result);

$resv_order_no = $result['resv_order_no'];

// 查询预约状态
$result = $sdyClient->getResv($resv_order_no);
echo "查询预约状态: \n";
var_dump($result);
```

更多接口的调用方式，请参见 example 或参看[官方文档](http://opendoc.sudiyi.cn/api.html)。

### 返回结果

 - GET、POST 类接口, 默认返回一个 Array 对象
 - PUT、DELETE 类接口, 若操作成功, 返回 Array 对象, 操作失败, 返回 NULL 或 Array
 - 请求错误, 抛出 SdyException 异常

```
预约成功: 
array(1) {
  'resv_order_no' =>
  int(20160719000001)
}

查询预约状态: 
array(3) {
  'status' =>
  int(1)
  'resv_order_no' =>
  int(20160719000001)
  'open_code' =>
  NULL
}
```

### 异常处理

SDK 执行过程中若遇到异常，将会抛出一个 SdyException 异常，用户可自行捕获并处理。

```php
use Sudiyi\Open\Api;
use Sudiyi\Open\Core\SdyException;

try {
    // 预约速递易箱格
    $result = $sdyClient->resv($data);
    $resv_order_no = $result['resv_order_no'];

    // 取消预约箱格
    $result = $sdyClient->cancelResv($resv_order_no);
    echo "取消预约箱格: \n";
    var_dump($result);
    

} catch (SdyException $e) {
    echo "============== ERROR ==============\n";
    echo $e->getMessage() . "\n";
    var_dump($e->getErrorBody());
    echo "===================================\n";
}
```

### 运行Sample程序

1. 修改 `example/api_use_demo.php`， 补充配置信息
2. 执行 `php ./example/api_use_demo.php`

### 运行单元测试

1. 执行`composer install`下载依赖的库
2. 设置环境变量

        export SDY_PARTNER_ID=partner_id
        export SDY_PARTNER_KEY=partner_key

3. 执行 `php vendor/bin/phpunit`

## 问题反馈

Goto: [ISSUES](https://github.com/sudiyi/sudiyi-open-php-sdk/issues)

## 开源协议

MIT





CHANGELOG.md
# ChangeLog - Sudiyi Open Platform PHP SDK

## v1.0.1 / 2016-07-19

* fix bug in `Api#cancelResv`
* finish phpunit tests

## v1.0.0 / 2016-07-14

* init repo


tests/ApiTest.php
<?php

namespace Sudiyi\Open\Tests;

use Sudiyi\Open\Api;

class ApiTest extends \PHPUnit_Framework_TestCase
{
    protected $api;
    protected $resv_order_no = 20160719000001;
    protected $device_id = 1000149;
    protected $area_id = 510109;
    
    protected function setUp()
    {
        $this->api = new Api(getenv('SDY_PARTNER_ID'), getenv('SDY_PARTNER_KEY'));
        
        $test_host = getenv('SDY_TEST_HOST');
        $test_host && $this->api->client->host = $test_host;
    }
    
    public function testResv()
    {
        $data = array(
            'device_id'         => $this->device_id,
            'box_type'          => 2,
            'notify_url'        => 'http://your.url/to/notify.php',
            'auto_upgd'         => true,
            'sender_name'       => 'Stefans',
            'sender_mobile'     => '18612345678',
            'order_no'          => '2015000001',
            'consignee_name'    => 'demon',
            'consignee_mobile'  => '15887654321',
        );
        
        $result = $this->api->resv($data);
        $this->assertArrayHasKey('resv_order_no', $result);
    }
    
    public function testGetResv()
    {
        $result = $this->api->getResv($this->resv_order_no);
        $this->assertArrayHasKey('resv_order_no', $result);
        $this->assertArrayHasKey('status', $result);
        $this->assertArrayHasKey('open_code', $result);
    }

    public function testGetResvStatus()
    {
        $results = $this->api->getResvStatus($this->resv_order_no);
        $result = array_pop($results);
        $this->assertArrayHasKey('resv_order_no', $result);
        $this->assertArrayHasKey('status', $result);
        $this->assertArrayHasKey('create_at', $result);
        $this->assertArrayHasKey('order_no', $result);
        $this->assertArrayHasKey('send_mobile', $result);
        $this->assertArrayHasKey('consignee_mobile', $result);
    }


    public function testCancelResv()
    {
        $result = $this->api->cancelResv($this->resv_order_no);
        $this->assertNull($result);
    }
    
    public function testGetArea()
    {
        $this->api->getArea();
    }
    
    public function testGetLattice()
    {
        $result = $this->api->getLattice($this->area_id);
        $this->assertEquals('array', gettype($result));
        $this->assertArrayHasKey('devices', $result[0]);
    }

    public function testGetBoxStatus()
    {
        $result = $this->api->getBoxStatus('device', $this->device_id);
        $this->assertArrayHasKey('small', $result);
        $this->assertArrayHasKey('medium', $result);
        $this->assertArrayHasKey('big', $result);
    }

    public function testGetOnlineBoxStatus()
    {
        $result = $this->api->getOnlineBoxStatus('device', $this->device_id);
        $this->assertEquals('array', gettype($result));
        $this->assertArrayHasKey('device', $result[0]);
    }

    public function testGetDeadLetter()
    {
        $result = $this->api->getDeadLetter();
        $this->assertArrayHasKey('post', $result);
        $this->assertArrayHasKey('put', $result);
    }

    public function testGetClosest()
    {
        $lat = 116.403119;
        $lng = 39.915378;
        $result = $this->api->getClosest($lat, $lng);
        $this->assertEquals('array', gettype($result));
    }

}


tests/BaseTest.php
<?php

namespace Sudiyi\Open\Tests;

use Sudiyi\Open\Core\Base;
use Sudiyi\Open\Core\Config;

class BaseTest extends \PHPUnit_Framework_TestCase
{
    
    public $partner_id = '1';
    public $partner_key = 'key';
    
    public $valid_sign = <<<BBB
POST
MTA0N2MyZmY0YzYwMThkMTY5MzY0MjNlY2EyYzZiZmU=
application/json;charset=UTF-8
Thu, 14 Jul 2016 21:51:28 +0800
v1/resv
BBB;

    protected $base;
    protected $path;
    protected $method;
    protected $content_type;
    protected $date;

    protected function setUp()
    {
        $this->base = new Base($this->partner_id, $this->partner_key);
        $this->path = 'v1/' . Config::RESV_URL;
        $this->method = 'post';
        $this->content_type = $this->base->content_type;
        $this->date = 'Thu, 14 Jul 2016 21:51:28 +0800';
    }

    public function testMakeContentMD5()
    {
        $result = $this->base->makeContentMD5('myContent');
        $this->assertEquals('MTA0N2MyZmY0YzYwMThkMTY5MzY0MjNlY2EyYzZiZmU=', $result);
        
        return $result;
    }

    /**
     * @depends clone testMakeContentMD5
     */
    public function testMakeSign($content_md5)
    {
        $result = $this->base->makeSign($this->method, $content_md5, $this->content_type, $this->date, $this->path);
        $this->assertEquals($this->valid_sign, $result);

        return $result;
    }

    /**
     * @depends clone testMakeSign
     */
    public function testMakeAuthorization($info2sign)
    {
        $result = $this->base->makeAuthorization($info2sign);
        $this->assertEquals('SDY 1:AoWdzFUcPilsC+xWU8D4ZhL0qxQ=', $result);
        
        return $result;
    }
    
    public function testGetHeader()
    {
        $result = $this->base->getHeader($this->path, $this->method);
        $this->assertCount(4, $result);
        $this->assertEquals([0,1,2,3], array_keys($result));
    }

}


tests/HttpTest.php
<?php

namespace Sudiyi\Open\Tests;

use Sudiyi\Open\Core\Http;

class HttpTest extends \PHPUnit_Framework_TestCase
{
    protected $http;
    
    public function testGet()
    {
        $this->http = new Http('http://sposter.net', 5);
        $result = $this->http->get('proxy/getnum', array('Content-type: application/json'));
        $this->assertArrayHasKey('result', $result);
        $this->assertArrayHasKey('packages_count', $result);
    }
    
    public function testGetInvalidJson()
    {
        $this->http = new Http('http://sposter.net', 5);
        $result = $this->http->get('/', []);
        $this->assertNull($result);
    }

    /** @noinspection PhpUndefinedNamespaceInspection */
    /** @noinspection PhpUndefinedClassInspection */
    /**
     * @expectedException Sudiyi\Open\Core\SdyException
     */
    public function testGetWithSdyException()
    {
        $this->http = new Http('http://bod.sudiyi.cn', 5);
        $this->http->get('404', []);
    }
    
    public function testPost()
    {
        $this->http = new Http('http://sposter.net', 5);
        $result = $this->http->post('/', [], ['post' => 'nothing']);
        $this->assertNull($result);
    }

    public function testDelete()
    {
        $this->http = new Http('http://sposter.net', 5);
        $result = $this->http->delete('/', []);
        $this->assertNull($result);
    }

    public function testPut()
    {
        $this->http = new Http('http://sposter.net', 5);
        $result = $this->http->put('/', [], ['put' => 'nothing']);
        $this->assertNull($result);
    }
        


}


src/Api.php
<?php

namespace Sudiyi\Open;

use Sudiyi\Open\Core\Base;
use Sudiyi\Open\Core\Config;
use Sudiyi\Open\Core\SdyException;
use Sudiyi\Open\Core\Http;

/**
 * Class Api
 * 
 * 接口访问类，包含所有对速递易开放平台API列表的封装
 * 
 * @package Sudiyi\Open
 */
class Api
{

    /**
     * @var string 接口版本号
     */
    public $version;

    /**
     * 构造函数
     *
     * @param String $partner_id  合作商家号,5位数字（开户邮件中可查看）
     * @param String $partner_key 合作商家密匙（必须配置，开户邮件中可查看）
     * @param String $version     接口版本号
     * @param int    $time_out    接口超时时间(单位:秒, 默认5秒)
     * @throws SdyException
     */
    public function __construct($partner_id, $partner_key, $version = 'v1', $time_out = 5)
    {
        $partner_id = trim($partner_id);
        $partner_key = trim($partner_key);
        $this->version = trim($version);

        if (! preg_match('/^\d{2,5}$/i', $partner_id))
            throw new SdyException("partner_id is invaild");

        if (empty($partner_key))
            throw new SdyException("partner_key is empty");

        if (! preg_match('/^v\d+$/i', $this->version))
            throw new SdyException("api version is wrong");

        $this->base = new Base($partner_id, $partner_key);
        $this->client = new Http(Config::HOST, $time_out);
    }

    /**
     * 预约箱格
     *
     * @param array $data  提交的数据
     * @return mixed
     */
    public function resv($data)
    {
        $path = $this->version . '/' . Config::RESV_URL;
        $header = $this->base->getHeader($path, 'post');

        return $this->client->post($path, $header, $data);
    }

    /**
     * 查看预约信息
     *
     * @param string $resv_order_no 订单号
     * @return mixed
     */
    public function  getResv($resv_order_no)
    {
        $path = $this->version . '/' . Config::RESV_URL . '/' . $resv_order_no;
        $header = $this->base->getHeader($path, 'get');

        return $this->client->get($path, $header);
    }

    /**
     * 取消预约箱格
     *
     * @param string $resv_order_no 订单号
     * @return mixed
     */
    public function cancelResv($resv_order_no)
    {
        $path = $this->version . '/' . Config::RESV_URL . '/' . $resv_order_no;
        $header = $this->base->getHeader($path, 'delete');
        
        return $this->client->delete($path, $header);
    }

    /**
     * 获取行政区域
     *
     * @return mixed
     */
    public function getArea()
    {
        $path = $this->version . '/' . Config::AREA_URL;
        $header = $this->base->getHeader($path, 'get');

        return $this->client->get($path, $header);
    }

    /**
     * 获取网点和设备信息
     *
     * @param integer $area_id 行政区域ID
     * @return mixed
     */
    public function getLattice($area_id)
    {
        $path = $this->version . '/' . Config::LATTICE_URL;
        $params = ['area' => $area_id];
        $header = $this->base->getHeader($path, 'get');

        return $this->client->get($path, $header, $params);
        
    }

    /**
     * 获取箱格状态,type为device,按照设备id获取，type为lattice,按照网点id获取
     *
     * @param integer $id   设备ID
     * @param string  $type 设备类型
     * @return mixed|void
     * @throws SdyException
     */
    public function getBoxStatus($type, $id)
    {
        if (! in_array($type, ['device', 'lattice']))
            throw new SdyException('Unknown $type was given in function getBoxStatus');

        $path = $this->version . '/' . Config::BOX_STATUS_URL;
        $params = [$type => $id];
        $header = $this->base->getHeader($path, 'get');

        return $this->client->get($path, $header, $params);
    }

    /**
     * 获取箱格状态和设备在线状态,type为device,按照设备id获取，type为lattice,按照网点id获取
     *
     * @param integer $id   设备ID
     * @param string  $type 设备类型
     * @return mixed
     * @throws SdyException
     */
    public function getOnlineBoxStatus($type, $id)
    {
        if (! in_array($type, ['device', 'lattice']))
            throw new SdyException('Unknown $type was given in function getOnlineBoxStatus');

        $path = $this->version . '/' . Config::ONLINE_BOX_STATUS_URL;
        $params = [$type => $id];
        $header = $this->base->getHeader($path, 'get');

        return $this->client->get($path, $header, $params);
    }

    /**
     * 获取死信
     *
     * @return mixed
     */
    public function getDeadLetter(){
        $path = $this->version . '/' . Config::DEAD_LETTER_URL;
        $header = $this->base->getHeader($path, 'get');

        return $this->client->get($path, $header);
    }

    /**
     * 查询订单当前的详细状态
     *
     * @param string $resv_order_no 订单号
     * @return mixed
     */
    public function getResvStatus($resv_order_no)
    {
        $path = $this->version . '/' . Config::RESV_URL;
        $params = ['resv_order_no' => $resv_order_no];
        $header = $this->base->getHeader($path, 'get');

        return $this->client->get($path, $header, $params);
    }

    /**
     * 获取附近设备
     *
     * @param string $lat 经度
     * @param string $lng 纬度
     * @return mixed
     */
    public function getClosest($lat, $lng)
    {
        $path = $this->version . '/' . Config::CLOSEST_URL;
        $params = ['lat' => $lat, 'lng' => $lng];
        $header = $this->base->getHeader($path, 'get');

        return $this->client->get($path, $header, $params);
    }
}



autoload.php
<?php

/**
 * An example of a project-specific implementation.
 *
 * After registering this autoload function with SPL, the following line
 * would cause the function to attempt to load the \Foo\Bar\Baz\Qux class
 * from /path/to/project/src/Baz/Qux.php:
 *
 *      new \Foo\Bar\Baz\Qux;
 *
 * @param string $class The fully-qualified class name.
 * @return void
 */
spl_autoload_register(function ($class) {

    // project-specific namespace prefix
    $prefix = 'Sudiyi\\Open\\';

    // base directory for the namespace prefix
    $base_dir = __DIR__ . '/src/';

    // does the class use the namespace prefix?
    $len = strlen($prefix);
    if (strncmp($prefix, $class, $len) !== 0) {
        // no, move to the next registered autoloader
        return;
    }

    // get the relative class name
    $relative_class = substr($class, $len);

    // replace the namespace prefix with the base directory, replace namespace
    // separators with directory separators in the relative class name, append
    // with .php
    $file = $base_dir . str_replace('\\', '/', $relative_class) . '.php';

    // if the file exists, require it
    if (file_exists($file)) {
        require $file;
    }
});


src/Core/Base.php
<?php

namespace Sudiyi\Open\Core;

/**
 * Class Base
 *
 * 基础类，该类中定义数据类最基本的行为，包括：
 * 设置签名、生成MD5、生成Authorization、设置http请求头部信息等
 *
 * @package Sudiyi\Open\Core
 */
class Base
{
    /**
     * @var Integer $partner_id  合作商家号
     */
    protected $partner_id;

    /**
     * @var String $partner_key  合作商家密匙
     */
    protected $partner_key;

    /**
     * @var String $content_type  Http Request MimeType.
     */
    public $content_type = Config::CONTENT_TYPE;

    /**
     * Base constructor.
     * 
     * @param $partner_id
     * @param $partner_key
     */
    public function __construct($partner_id, $partner_key)
    {
        $this->partner_id = $partner_id;
        $this->partner_key = $partner_key;
        date_default_timezone_set('Asia/Shanghai');
    }

    /**
     * 生成MD5
     *
     * @param  string $str 需要生成MD5的字符串
     * @return string
     */
    public function makeContentMD5($str)
    {
        return base64_encode(md5($str));
    }

    /**
     * 生成签名
     *
     * @param string $method
     * @param string $content_md5
     * @param string $content_type
     * @param string $date
     * @param string $path
     * @return string
     */
    public function makeSign($method, $content_md5, $content_type, $date, $path)
    {
        return join("\n", [strtoupper($method), $content_md5, $content_type, $date, $path]);
    }

    /**
     * 生成Authorization
     * @param string $info2sign  签名
     * @return string
     */
    public function makeAuthorization($info2sign)
    {
        $signature = base64_encode(hash_hmac('sha1', $info2sign, $this->partner_key, true));

        return "SDY {$this->partner_id}:{$signature}";
    }

    /**
     * 设置HTTP请求的头部信息
     *
     * @param string  $path    HTTP-Request-Path
     * @param string  $method  HTTP-Request-Method
     * @return array
     */
    public function getHeader($path, $method='get')
    {
        $date = date('r');
        $content_md5 = $this->makeContentMD5('');
        $path = '/' . ltrim($path, '/');
        
        $info2sign = $this->makeSign($method, $content_md5, $this->content_type, $date, $path);
        $authorization = $this->makeAuthorization($info2sign);

        $header = array(
            'Content-Type:' . $this->content_type,
            'Date:' . $date,
            'Content-MD5:' . $content_md5,
            'Authorization:' . $authorization,
        );

        return $header;
    }

}


src/Core/Http.php
<?php

namespace Sudiyi\Open\Core;

/**
 * Class Http
 *
 * Http 请求处理
 *
 * @package Sudiyi\Open\Core
 */
class Http
{
    /**
     * @var string  Sudiyi Open Platform API Host.
     */
    public $host;

    /**
     * @var int     Request timeout,
     */
    public $time_out = 5;

    /**
     * Http constructor.
     *
     * @param string  $host
     * @param integer $time_out
     * @throws SdyException
     */
    public function __construct($host, $time_out = null)
    {
        if (substr($host, 0, 4) != 'http')
            throw new SdyException('A illegal host was given when Http::new.');
        else
            $this->host = rtrim($host, '/');

        isset($time_out) && $this->time_out = $time_out;
    }
    
    /**
     * curl
     *
     * @param string $url
     * @param mixed  $header
     * @param string $method
     * @param null   $post_data
     * @return mixed
     * @throws SdyException
     */
    protected function curl($url, $header, $method, $post_data = null)
    {
        $ch = curl_init();
        //设置超时
        curl_setopt($ch, CURLOPT_TIMEOUT, $this->time_out);
        //设置header
        curl_setopt($ch, CURLOPT_HTTPHEADER, $header);
        curl_setopt($ch, CURLOPT_HEADER, 0);
        curl_setopt($ch, CURLOPT_FOLLOWLOCATION, 0);
        //要求结果为字符串且输出到屏幕上
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
        curl_setopt($ch, CURLOPT_URL, $url);
        switch (strtoupper($method)) {
            case 'GET' :
                curl_setopt($ch, CURLOPT_HTTPGET, 1);
                break;
            case 'POST':
                curl_setopt($ch, CURLOPT_POST, 1);
                curl_setopt($ch, CURLOPT_POSTFIELDS, $post_data);
                break;
            case 'PUT' :
                curl_setopt($ch, CURLOPT_CUSTOMREQUEST, 'PUT');
                curl_setopt($ch, CURLOPT_POSTFIELDS, $post_data);
                break;
            case 'DELETE':
                curl_setopt($ch, CURLOPT_CUSTOMREQUEST, 'DELETE');
                curl_setopt($ch, CURLOPT_POSTFIELDS, $post_data);
                break;
        }
        curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);
        curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, false);

        $response = curl_exec($ch);
        $info = curl_getinfo($ch);
        curl_close($ch);

        if (in_array($info['http_code'], [200, 204])) {

            return json_decode($response, true);

        } else {
            $error_details = array(
                'url' => $url,
                'status' => $info['http_code'],
                'response' => substr($response, 0, 140),
                'body' => array(
                    'url' => $url,
                    'header' => $header,
                    'method' => $method,
                    'post_data' => $post_data,
                )
            );

            throw new SdyException($error_details);
        }
    }

    /**
     * HTTP Get.
     *
     * @param string     $path      Request path.
     * @param string     $header    Http header which is created by Base#getHeader.
     * @param array|null $params    Query array of the request.
     * @return mixed
     * @throws SdyException
     */
    public function get($path, $header, array $params = null)
    {
        $url = $this->host . '/' . $path;
        isset($params) && $url .= '?' . http_build_query($params);
        return $this->curl($url, $header, 'get');
    }

    /**
     * Http Post.
     *
     * @param string $path      Request path.
     * @param string $header    Http header which is created by Base#getHeader.
     * @param mixed  $data      Request body content.
     * @return mixed
     * @throws SdyException
     */
    public function post($path, $header, $data)
    {
        $url = $this->host . '/' . $path;
        return $this->curl($url, $header, 'post', json_encode($data));
    }

    /**
     * Http Put.
     *
     * @param string $path      Request path.
     * @param string $header    Http header which is created by Base#getHeader.
     * @param mixed  $data      Request body content.
     * @return mixed
     * @throws SdyException
     */
    public function put($path, $header, $data)
    {
        $url = $this->host . '/' . $path;
        return $this->curl($url, $header, 'put', json_encode($data));
    }

    /**
     * HTTP Delete.
     *
     * @param  string  $path    Request path.
     * @param  string  $header  Http header which is created by Base#getHeader.
     * @return mixed
     * @throws SdyException
     */
    public function delete($path, $header)
    {
        $url = $this->host . '/' . $path;
        return $this->curl($url, $header, 'delete');
    }

}

src/Core/Config.php
<?php

namespace Sudiyi\Open\Core;

/**
 * Class Config
 *
 * 基础配置信息
 *
 * @package Sudiyi\Open\Core
 */
class Config
{
    //=======【HTTP请求头部信息设置】=====================================
    //
    /**
     *
     * CONTENT_TYPE: http请求类型
     *
     */
    const CONTENT_TYPE = 'application/json;charset=UTF-8';

    //=======【开放平台接口地址设置】=====================================
    //
    /**
     *
     * HOST：速递易开放平台接口域名
     * AREA_URL：获取行政区域接口地址
     * LATTICE_URL：获取网点和设备信息接口地址
     * BOX_STATUS_URL：获取箱格状态接口地址
     * ONLINE_BOX_STATUS_URL：获取箱格状态和设备在线状态接口地址
     * RESV_URL:预约箱格接口地址
     * DEAD_LETTER_URL：获取死信接口地址
     * CLOSEST_URL：获取附近设备接口地址
     *
     */
    const HOST = 'http://open.sudiyi.cn';
    const AREA_URL = 'area';
    const LATTICE_URL = 'lattice';
    const BOX_STATUS_URL = 'boxStatus';
    const ONLINE_BOX_STATUS_URL = 'boxStatus/online';
    const RESV_URL = 'resv';
    const DEAD_LETTER_URL = 'deadletter';
    const CLOSEST_URL = 'closest';
}


example/notify_demo.php
<?php

/**
 *  回调基础类
 * @author yhs
 */
$post = json_decode(file_get_contents("php://input"), true);


/**
 * 速递易返回分为两种，一种是有result的（预约完箱格之后一般几秒会调用），一个是有status的（箱格有变化的时候就会通知）。所以要分别来判断；
 */

//返回有result表示预约之后返回的
if (isset($post['result'])) {
    $resv_order_no = $post['resv_order_no'];//预约订单的编号
    $result = $post['result'];//预约结果*
    $order_no = $post['order_no'];//第三方订单,为发起预约时传入的订单号
    $box_type = $post['box_type'];//实际取得的箱格的类型
    $box_no = $post['box_no'];//箱格序号
    $upgraded = $post['upgraded'];//是否进行了升箱
    $resv_code = $post['resv_code'];//预约码

    if ($result == 'success') {
        //TODO 如果预约成功
    } elseif ($result == 'fail') {
        //TODO 如果预约失败
    } elseif ($result == 'timeout') {
        //TODO 如果预约超时
    }
}

//返回有status的表示箱格状态有变化的
if (isset($post['status'])) {
    $resv_order_no = $post['resv_order_no'];//预约订单的编号
    $status = $post['status'];//订单的状态码
    $open_code = $post['open_code'];//开箱码
    
    if ($status == 5){
        //TODO 已投件，等待客户取件
    } elseif ($status == 6){
        //TODO 已取件，通知给相应人员
    } elseif ($status == 7){
        //TODO 客户预约超过保留时长仍未投件，预约箱格自动释放
    }
}


src/Core/SdyException.php
<?php

namespace Sudiyi\Open\Core;

/**
 * Class SdyException
 *
 * 异常类
 *
 * @package Sudiyi\Open\Core
 */
class SdyException extends \Exception
{
    private $details = array();

    function __construct($details)
    {
        if (is_array($details)) {
            $this->details = $details;
            array_pop($details);
            $message = json_encode($details, 320);
            parent::__construct($message);
        } else {
            $message = $details;
            parent::__construct($message);
        }
    }

    public function getHTTPStatus()
    {
        return isset($this->details['status']) ? $this->details['status'] : '';
    }

    public function getErrorCode()
    {
        return isset($this->details['code']) ? $this->details['code'] : '';
    }

    public function getErrorResponse()
    {
        return isset($this->details['response']) ? $this->details['response'] : '';
    }

    public function getErrorBody()
    {
        return isset($this->details['body']) ? $this->details['body'] : '';
    }
}


example/api_use_demo.php
<?php
/**
 * Api use demo.
 */

require_once  __DIR__ . '/../autoload.php';

use Sudiyi\Open\Api;
use Sudiyi\Open\Core\SdyException;

/*
 * 合作商家号、合作商家密匙（开户邮件中可查看）
 * 本 Demo 使用速递易开放平台测试环境 (via setting $sdyObj->client->host.)
 */
$partner_id = '<您从速递易开放平台获得的 PartnerId>';
$partner_key = '<您从速递易开放平台获得的 PartnerKey>';

/* 尝试从环境变量中获取配置 */
if (intval($partner_id) == 0) {
    $partner_id = getenv('SDY_PARTNER_ID');
    $partner_key = getenv('SDY_PARTNER_KEY');
}

/* 初始化 Api 实例 */
$sdyObj = new Api($partner_id, $partner_key);

/* 修改测试服务器 Host (如果有配置) */
$test_host = getenv('SDY_TEST_HOST');
$test_host && $sdyObj->client->host = $test_host;


$area_id = 510109;
$device_id = 1000149;
$lat = 116.403119;
$lng = 39.915378;

$data = array(
    'device_id'         => $device_id,                      // 速递易设备ID
    'box_type'          => 2,                               // 箱格类型 0:大箱 1:中箱 2:小箱 3:冰箱
    'notify_url'        => 'http://your.url/to/notify.php', // 回调第三方系统的地址
    'auto_upgd'         => true,                            // 选填，是否可以自动升箱
    'sender_name'       => 'Stefans',                       // 选填，投递人姓名
    'sender_mobile'     => '18612345678',                   // 选填，投递人手机
    'order_no'          => '2015000001',                    // 第三方订单号,即自己平台的订单号
    'consignee_name'    => 'demon',                         // 收货人姓名
    'consignee_mobile'  => '15887654321',                   // 收货人手机号
    'payment'           => 0,                               // 选填，到付金额，如果不是到付件则为0
    'pay_type'          => 0,                               // 选填，超期付费方式 0:快递柜付费 1:从商户账户扣除
    'duration'          => 60                               // 选填，预约时长，单位分钟，默认240分钟
);

try {
    // 预约速递易箱格
    $result = $sdyObj->resv($data);
    echo "预约成功: \n";
    var_dump($result);

    $resv_order_no = $result['resv_order_no'];

    // 查询预约状态
    $result = $sdyObj->getResv($resv_order_no);
    echo "查询预约状态: \n";
    var_dump($result);

    // 查询订单当前的详细状态
    $result = $sdyObj->getResvStatus($resv_order_no);
    echo "查询订单当前的详细状态: \n";
    var_dump($result);

    // 取消预约箱格
    
    $result = $sdyObj->cancelResv($resv_order_no);
    echo "取消预约箱格: \n";
    if ($result == null)
        echo"此次预约已不能取消 \n";
    else
        var_dump($result);
        
    
    // 获取速递易的行政区域,查出来成都市的id为:510100,成都市高新区的id为：510109,其他城市根据自己需要来
    $result = $sdyObj->getArea();
    echo "获取行政区域: \n";
    var_dump($result);

    // 获取网点和设备信息,根据上面查出来的id进行查询，会查询出来该行政区的所有设备信息，比如查询出来 SDY成都高新区-世纪城路-龙湖世纪峰景B50 的设备id为 1001412，则可以预约了
    $result = $sdyObj->getLattice($area_id);
    echo "获取网点和设备信息: \n";
    var_dump($result);

    // 获取箱格状态,type为device,按照设备id获取，type为lattice,按照网点id获取
    $result = $sdyObj->getBoxStatus('device', $device_id);
    echo "获取箱格状态: \n";
    var_dump($result);

    // 获取箱格状态和设备在线状态,type为device,按照设备id获取，type为lattice,按照网点id获取
    $result = $sdyObj->getOnlineBoxStatus('device', $device_id);
    echo "获取箱格状态和设备在线状态: \n";
    var_dump($result);

    // 获取死信
    $result = $sdyObj->getDeadLetter();
    echo "获取死信: \n";
    var_dump($result);

    // 获取附近设备
    $result = $sdyObj->getClosest($lat, $lng);
    echo "获取附近设备: \n";
    var_dump($result);

} catch (SdyException $e) {
    echo "============== ERROR ==============\n";
    echo $e->getMessage() . "\n";
    var_dump($e->getErrorBody());
    echo "===================================\n";
}


